src/query.rs

```diff
@@ -30,18 +32,28 @@ pub unsafe trait Fetch<'a>: Sized {
     /// Type of value to be fetched
     type Item;
 
+    /// The type of the data which can be cached to speed up retrieving
+    /// the relevant type states from a matching [archetype][Archetype]
+    type State: Copy;
+
     /// A value on which `get` may never be called
     fn dangling() -> Self;
 
     /// How this query will access `archetype`, if at all
     fn access(archetype: &Archetype) -> Option<Access>;
 
     /// Acquire dynamic borrows from `archetype`
-    fn borrow(archetype: &Archetype);
+    fn borrow(archetype: &Archetype, state: Self::State);
     /// Construct a `Fetch` for `archetype` if it should be traversed
-    fn new(archetype: &'a Archetype) -> Option<Self>;
+    fn new(archetype: &'a Archetype) -> Option<Self> {
+        Self::prepare(archetype).map(|state| Self::execute(archetype, state))
+    }
```

Comment:
It looks like you've mostly refactored uses of this into the new form. Can it be removed entirely?

Reply:
I have added another commit that open codes this instead. Please have a look if you prefer the result and I will squash it. Otherwise, I can drop the commit.

Reply:
Net reduction in LoC, one less trait method, LGTM.